r/R/filesystem.R

```diff
@@ -273,33 +273,53 @@ FileSystem$from_uri <- function(uri) {
   fs___FileSystemFromUri(uri)
 }
 
-get_path_and_filesystem <- function(x, filesystem = NULL) {
+get_paths_and_filesystem <- function(x, filesystem = NULL) {
   # Wrapper around FileSystem$from_uri that handles local paths
   # and an optional explicit filesystem
   if (inherits(x, "SubTreeFileSystem")) {
     return(list(fs = x$base_fs, path = x$base_path))
   }
-  assert_that(is.string(x))
-  if (is_url(x)) {
+  assert_that(is.character(x))
+  are_urls <- are_urls(x)
+  if (any(are_urls)) {
+    if (!all(are_urls)) {
+      stop("Vectors of mixed paths and URIs are not supported", call. = FALSE)
```

Comment:
Why not? I wonder if it's not the concern of this utility function whether it gets paths, URIs, SubTreeFileSystems, etc.

Reply:
The function in the C++ library that this ultimately calls to generate a dataset from multiple files expects all those files to be in the same filesystem.

I suppose in the case where the user mixed `file://` URIs and plain file paths with no protocol, we could recognize that they both point to the local file system, but that would require extra code and I I don't think it would add any practical value

Reply:
My point was just that this utility knows nothing about datasets or any of that, it just takes a reference to a file (or in this case now, a vector of them), which may be a string path (with an optional `filesystem` specified), a URI, or a SubTreeFileSystem (a hacky way to specify a fs+path together), and returns a standard (fs, path) tuple. It seems like the responsibility of FileSystemDatasetFactory or whatever to validate that the resulting fs+paths are conforming to its requirements.

Reply:
Ah, ok. The reason we need to assert here that these are all URIs or all file paths is that the code forks here based on this, and the case in which some are URIs and some are file paths is unhandled and results in unhelpful errors